Add Appveyor for Windows CI

tgalopin opened the pull request:
Related to https://github.com/puli/issues/issues/115.

stof commented:
You need the curl extension for this repo

stof commented:
I suggest adding the ``files`` directory of the Composer cache in the AppVeyor cache, to be able to reuse the same downloaded archives between builds, making it much less likely to hit the Github API rate limit, and making the composer installation faster.
I don't remember where Composer puts its cache on Windows though. You have 2 solutions:

- figuring out where it is located
- setting the ``COMPOSER_CACHE_DIR`` env variable, to control yourselves where the composer cache is stored.

This applies to all repos of course

tgalopin commented:
I added the Composer cache (according to the documentation: https://getcomposer.org/doc/03-cli.md#composer-cache-dir) and I fixed the cURL issue. There still are failures, but normal ones than can have their own PR IMO.

stof commented on appveyor.yml:
```diff
@@ -0,0 +1,35 @@
+clone_folder: c:\projects\puli\manager
+
+cache:
+  - '%LOCALAPPDATA%\Composer'
```
you should cache only the ``files`` subfolder between builds. Here is how the composer cache is structured:

- ``files``: cache for downloaded archives
- ``vcs``: cache for cloned repositories when registering VCS repository in which metadata should be looked at (and when they are not on Packagist or Bitbucket as Composer can use the API rather than cloning in such cases). This is not used here as all deps are coming from Packagist
- ``repo``: cache for metadata files downloaded from composer repositories (only Packagist in the case of puli)

Caching the ``repo`` folder between builds will do more harm than good:

- Packagist metadata will be updated between each Puli build (unless you are really very fast at triggering multiple builds, as Packagist can potentially update them every minute)
- Puli depends on active repositories (PHPUnit, Symfony, etc...), which means that even the per-package files are unlikely to be reused when there is several days between builds.

This means that the cache will miss most of the time. The time won caching these files will probably be very small here.

On the other hand, caching the ``files`` folder is beneficial:

- new files are added in it only when the version selected for a package changes, i.e. only when one of your dependencies makes a new release. This happens far less often than when one of them makes a new commit in their repo
- archives are bigger than JSON files from Packagist, and getting them from the cache saves time
- downloads are counted in the Github API rate limit. Making a download only in case of a cache miss means you will make such downloads far less often. And the API rate limit is 60 requests per IP per hour for anonymous access, which could be reached quickly (PHPUnit alone brings at least 15 packages, and so 15 downloads. And this does not cound the actual dependencies of the puli library).

stof replied:
btw, this should be changed on Travis too.

tgalopin replied:
Thanks for your explainations, I didn't know that. I'll add the change for Travis in this PR.

stof replied:
Well, I figured the best caching strategy for Travis only after some trial.

A PR on the repository package is needed to add the cache btw.

webmozart replied:
Thank you for the detailed explanation @stof! :) Much appreciated!

webmozart commented:
Thanks a lot for your work the two of you!! :)
